foxpuppet/ui/browser/window.py

```diff
@@ -0,0 +1,90 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+
+from .tabbar import TabBar
+from .navbar import Navbar
+from foxpuppet.ui.windows import Windows
+from foxpuppet.ui.base_window import BaseWindow
+
+
+class Browser_Window(object):
+
+    def __init__(self, foxpuppet, *args, **kwargs):
+        BaseWindow()
+        self.windows = Windows(foxpuppet)
+
+        self.selenium = foxpuppet.selenium
+        self._navbar = None
+        self._tabbar = None
+
+    @property
+    def navbar(self):
+        self.selenium.switch_to()
+
+        if not self._navbar:
+            navbar = self.selenium.find_element_by_id('nav-bar')
+            self._navbar = Navbar(lambda: self.selenium, self, navbar)
+
+        return self._navbar
+
+    @property
+    def tabbar(self):
+        self.selenium.switch_to.window(self.windows.current)
+
+        if not self._tabbar:
+            tabbrowser = self.selenium.find_element_by_id('tabbrowser-tabs')
+            self._tabbar = TabBar(lambda: self.selenium, self, tabbrowser)
+
+        return self._tabbar
+
+    def close(self, handle):
+        self.selenium.switch_to.window(handle)
+        self.selenium.find_element_by_id('titlebar-close').click()
```

Comment:
Separating out element retrieval and actions? It would give re-usable code, in case a test only wants to have the close button element.